File: 4I505 - CPA/Project/src/C/circle.c
#include "circle.h"

struct cl naive(struct pt points[]) {
  size_t size = pts_size(points);
  struct cl circle = circle_nil();

  for (size_t i = 0; i < size; i++)
    for (size_t j = 0; j < size; j++) {
      struct pt p1 = points[i], p2 = points[j];
      if (p1.equal(p1, p2))
        continue;
      bool all_covered = true;
      struct cl temp = new_circle(middle  (p1, p2),
                                  distance(p1, p2) / 2);
      for (size_t k = 0; k < size; k++)
        if (! temp.cover(temp, points[k])) {
          all_covered = false;
          break;
        }
      if (all_covered)
        circle = temp;
    }
  if (! circle.equal(circle, circle_nil()))
    return circle;

  for (size_t i = 0; i < size; i++)
    for (size_t j = 0; j < size; j++) {
      struct pt p1 = points[i], p2 = points[j];
      if (p1.equal(p1, p2))
        continue;
      for (size_t k = 0; k < size; k++) {
        struct pt p3 = points[k];
        if (p3.equal(p3, p1) ||
            p3.equal(p3, p2))
          continue;
        bool all_covered = true;
        struct cl temp = circum_circle(p1, p2, p3);
        for (size_t l = 0; l < size; l++)
          if (! temp.cover(temp, points[l])) {
            all_covered = false;
            break;
          }
        if (all_covered)
          circle = temp;
      }
    }

  return circle;
}

File: 4I505 - CPA/Project/src/Ruby/circle.rb
require 'benchmark'
require_relative 'geometry'
require_relative 'reading'

def naive(points)
  @circle = Geometry.circleNil
  points.each do |q|
    points.each do |r|
      next if q == r
      all_covered = true
      temp = Geometry::Circle.new(Geometry.middle(q, r), Geometry.distance(q, r) / 2)
      points.each do |s|
        unless temp.is_covered s
          all_covered = false
          break
        end
      end

      if all_covered
        if temp.radius < @circle.radius or @circle == Geometry.circleNil
          @circle = temp
        end
      end
    end
  end
  return @circle unless @circle == Geometry.circleNil

  points.each do |q|
    points.each do |r|
      next if q == r
      points.each do |s|
        next if s == r or s == q
        all_covered = true
        temp = Geometry.circumcircle(q, r, s)
        points.each do |t|
          unless temp.is_covered t
            all_covered = false
            break
          end
        end

        if all_covered
          if temp.radius < @circle.radius or @circle == Geometry.circleNil
            @circle = temp
          end
        end
      end
    end
  end
  @circle
end

tests = Dir.entries(ARGV.first)
tests.shift
tests.shift
File.open "results_naive.csv", "w" do |fh|
  fh << "x;y;radius;time\n"
  tests.each do |t|
    $stdout << t.to_s << "\n"
    points = read_points t
    circle = nil
    bench = Benchmark.realtime { circle = naive points }
    fh << circle.center.x << ";" << circle.center.y << ";" << circle.radius << ";" << bench << "\n"
    fh.flush
  end
end
